File: chapter5/example/binary_search.c
typedef int DataType;
int n, t;
int * x;

int binarysearch(DataType t);

int binarysearch(DataType t) {
    int left, right, middle;
    left = 0;
    right = n - 1;

    while (left <= right) {
        middle = (left + right) / 2;
        /* printf("%d %d %d\n", left, middle, right); */
        if (x[middle] < t) left = middle + 1;
        else if (x[middle] == t) {
            return middle;
        }
        else right = middle - 1;
    }
    return -1;
}

File: chapter5/5_8/binary_search.rb
def binary_search(array, target)
  length = array.length
  left = 0
  right = length - 1
  while (left <= right) do
    middle = (left + right) / 2
    if (array[middle] > target)
      right = middle - 1
    elsif (array[middle] < target)
      left = middle + 1
    else
      return middle
    end
  end
  return -1
end
